torchdynamo/symbolic_convert.py

```diff
@@ -447,15 +447,21 @@ def import_source(self, module_name):
 
     def IMPORT_NAME(self, inst):
         level, fromlist = self.popn(2)
+        level = level.as_python_constant()
+        fromlist = fromlist.as_python_constant()
         module_name = inst.argval
         value = __import__(
             module_name,
-            fromlist=fromlist.as_python_constant(),
-            level=level.as_python_constant(),
+            fromlist=fromlist,
+            level=level,
         )
-        # __import__ returns the top level package name, so the source has to be
-        # set correctly.
-        if fromlist.as_python_constant() is None and hasattr(value, "__name__"):
+
+        # For __import__, when the name variable is of the form package.module,
+        # normally, the top-level package (the name up till the first dot) is
+        # returned, not the module named by module_name. However, when a
+        # non-empty fromlist argument is given, the module named by name is
+        # returned. Therefore, we set the source correctly here.
+        if fromlist is None and hasattr(value, "__name__"):
```

Comment:
The docs seem to imply this behavior also would occur if `fromlist` was an empty list? While unlikely, hypothetically someone could emit a bytecode that passes an empty list here rather than None?

Reply:
Regarding the hasattr, if we cannot guarantee that the module has the `__name__` attribute set accurately (and it *is* overrideable, e.g., `import b; b.__name__ = "arf"`, it would seem that it would be better to do this via `module_name`.  It is probably best to replicate the logic in CPython's `__import__` implementation:

```
    if not fromlist:
        # Return up to the first dot in 'name'. This is complicated by the fact
        # that 'name' may be relative.
        if level == 0:
            return _gcd_import(name.partition('.')[0])
        elif not name:
            return module
        else:
            # Figure out where to slice the module's name up to the first dot
            # in 'name'.
            cut_off = len(name) - len(name.partition('.')[0])
            # Slice end needs to be positive to alleviate need to special-case
            # when ``'.' not in name``.
            return sys.modules[module.__name__[:len(module.__name__)-cut_off]]
    elif hasattr(module, '__path__'):

```

Reply:
For the `fromlist` empty list, I don't think the Python bytecode can have empty list. I tried a couple of examples
~~~
import dis
def fn1():
    from math import sqrt
dis.dis(fn1)


def fn2():
    import math.sqrt
dis.dis(fn2)
~~~
~~~
  3           0 LOAD_CONST               1 (0)
              2 LOAD_CONST               2 (('sqrt',))
              4 IMPORT_NAME              0 (math)
              6 IMPORT_FROM              1 (sqrt)
              8 STORE_FAST               0 (sqrt)
             10 POP_TOP
             12 LOAD_CONST               0 (None)
             14 RETURN_VALUE

  8           0 LOAD_CONST               1 (0)
              2 LOAD_CONST               0 (None)
              4 IMPORT_NAME              0 (math.sqrt)
              6 STORE_FAST               0 (math)
              8 LOAD_CONST               0 (None)
             10 RETURN_VALUE
~~~


For the hasattr, I am going to replicate the logic. Thanks for passing the pointer.

Reply:
I mean, sure, I agree it's unlikely, but the tweak to handle empty list here is so slight (just do a bool-ish test directly on `fromlist`), there's no downside to doing it. (And you will match the test that `__import__` in cpython does too...)